app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt

```diff
@@ -1,59 +1,345 @@
-        TODO("Not yet implemented")
+        // create a space for the new question in sb and save its id
+        val newChildRef = db.child(questionsPath).push()
+        val questionId = newChildRef.key ?: error("Failed to generate question ID")
+        // create the new question using given parameters
+        val question = Question(questionId, course.courseId, user.userId, questionText ?: "", emptyList())
+        // add the new question in the db
+        newChildRef.setValue(question)
+
+        //add the question in the course's questions list
+        db.child(coursesPath).child(course.courseId).child(questionsPath).child(questionId).setValue(question)
+        getCourseQuestions(course)
+
+        //add the question in the user's questions list
+        db.child(usersPath).child(user.userId).child(questionsPath).child(questionId).setValue(question)
+
+        return question
     }
 
     override fun addAnswer(user: User, question: Question, answerText: String?): Answer {
-        TODO("Not yet implemented")
-    }
+        // create a space for the new answer in sb and save its id
+        val newChildRef = db.child(answersPath).push()
+        val answerId = newChildRef.key ?: error("Failed to generate answer ID")
+        // create the new answer using given parameters
+        val answer = Answer(answerId, question.questionId, user.userId, answerText ?: "")
+        newChildRef.setValue(answer)
+
+        //add the answer in the question's answers list
+        db.child(questionsPath).child(question.questionId).child(answersPath).child(answerId).setValue(answer)
+
+        //add the answer in the user's questions list
+        db.child(usersPath).child(user.userId).child(answersPath).child(answerId).setValue(answer)
+
+        //add the answer in the course's question's answers list
+        db.child(coursesPath).child(question.courseId).child(questionsPath).child(question.questionId).child(answersPath).child(answerId).setValue(answer)
 
-    override fun addUser(userId: String, username: String): User {
-        TODO("Not yet implemented")
+        //add the answer in the user's question's answers list
+        db.child(usersPath).child(user.userId).child(questionsPath).child(question.questionId).child(answersPath).child(answerId).setValue(answer)
+
+
+        return answer
     }
 
-    override fun getUserQuestions(user: User): Set<Question> {
-        TODO("Not yet implemented")
+    override fun addUser(userId:String, username: String): User {
+
+        val getUser = getUserById(userId)
+        if(getUser != null){
+            return getUser
+        }
+
+        // create a space for the new question in sb and save its id
+        val newUser = User(userId, username, emptyList(), emptyList())
+        db.child(usersPath).child(userId).setValue(newUser)
+        return newUser
     }
 
     override fun getQuestionById(id: String): Question? {
```

Comment:
Same remark here for the functions getQuestionById(id: String), getAnswerById(id: String), getUserById(id: String) and getCourseById(id: String)